diff --git a/server/devpi_server/model.py b/server/devpi_server/model.py
index f806a585..d052efe0 100644
--- a/server/devpi_server/model.py
+++ b/server/devpi_server/model.py
@@ -33,6 +33,7 @@ from typing import cast
 import functools
 import getpass
 import json
+import os
 import re
 import warnings
 
@@ -1122,6 +1123,14 @@ class BaseStage:
         # so existing indexes work as before
         whitelist_inheritance = self.get_whitelist_inheritance()
         whitelist = None
+        blacklist = os.environ.get('DEVPI_MIRROR_BLACKLIST', '')
+        if blacklist:
+            if os.path.isfile(blacklist):
+                with open(blacklist) as file:
+                    blacklist = file.readlines() #[x.strip() for x in file.readlines()]
+                    blacklist = [x for x in blacklist if x]
+            else:
+                blacklist = re.split(',|;|:', blacklist)
         for stage in self.sro():
             if stage.ixconfig["type"] == "mirror":
                 if private_hit:
@@ -1148,6 +1157,8 @@ class BaseStage:
                     whitelisted = stage
                 elif stage.has_project_perstage(project):
                     private_hit = True
+                elif blacklist and project in blacklist:
+                    private_hit = True
 
             try:
                 exists = stage.has_project_perstage(project)
