diff --git a/setup.py b/setup.py
index 9aff8d2..38b437c 100755
--- a/setup.py
+++ b/setup.py
@@ -136,7 +136,9 @@ def main():
             r"python/cv2/.*config.*.py"
         ]
         +
-        [ r"python/cv2/py.typed" ] if sys.version_info >= (3, 6) else []
+        # https://github.com/opencv/opencv-python/issues/871
+        # [ r"python/cv2/py.typed" ] if sys.version_info >= (3, 6) else []
+        []
         ,
         "cv2.data": [  # OPENCV_OTHER_INSTALL_PATH
             ("etc" if os.name == "nt" else "share/opencv4") + r"/haarcascades/.*\.xml"
@@ -155,8 +157,9 @@ def main():
         ],
     }
 
-    if sys.version_info >= (3, 6):
-        rearrange_cmake_output_data["cv2.typing"] = ["python/cv2" + r"/typing/.*\.py"]
+    # https://github.com/opencv/opencv-python/issues/871
+    # if sys.version_info >= (3, 6):
+    #    rearrange_cmake_output_data["cv2.typing"] = ["python/cv2" + r"/typing/.*\.py"]
 
     # Files in sourcetree outside package dir that should be copied to package.
     # Raw paths relative to sourcetree root.
@@ -502,8 +505,8 @@ def get_and_set_info(contrib, headless, rolling, ci_build):
     version_file = os.path.join(here, "cv2", "version.py")
 
     # generate a fresh version.py always when Git repository exists
-    # (in sdists the version.py file already exists)
-    if os.path.exists(".git"):
+    # prevent the submodules from being re-synced to their original commit
+    if False:  # os.path.exists(".git"):
         old_args = sys.argv.copy()
         sys.argv = ["", str(contrib), str(headless), str(rolling), str(ci_build)]
         runpy.run_path("find_version.py", run_name="__main__")
